src/main/gdk/SDL_sysmain_runapp.cpp

```diff
@@ -30,119 +30,160 @@ extern "C" {
-        } else {
-            SDL_SetError("[GDK] Unable to get titleid. Will not call XblInitialize. Check MicrosoftGame.config!");
+        char *argdata = ((char *)argv) + (argc + 1) * sizeof(*argv);
+        int argdata_index = 0;
+
+        for (int i = 0; i < argc; ++i) {
+            const int bytes_written = WideCharToMultiByte(CP_UTF8, 0, argvw[i], -1, argdata + argdata_index, argdata_size - argdata_index, NULL, NULL);
+            if (!bytes_written) {
+                result = ErrorProcessingCommandLine();
+                goto cleanup;
+            }
+            argv[i] = argdata + argdata_index;
+            argdata_index += bytes_written;
         }
+        argv[argc] = NULL;
 
-        SDL_SetMainReady();
-
-        if (!GDK_RegisterChangeNotifications()) {
-            return -1;
-        }
-
-        // Run the application main() code
         result = mainFunction(argc, argv);
 
-        GDK_UnregisterChangeNotifications();
+    cleanup:
+        HeapFree(GetProcessHeap(), 0, argv);
+        LocalFree(argvw);
+    }
+
+    return result;
+}
 
-        // !!! FIXME: This follows the docs exactly, but for some reason still leaks handles on exit?
-        // Terminate the task queue and dispatch any pending tasks
-        XTaskQueueTerminate(taskQueue, false, nullptr, nullptr);
-        while (XTaskQueueDispatch(taskQueue, XTaskQueuePort::Completion, 0))
-            ;
+/* Gets the arguments with GetCommandLine, converts them to argc and argv
+   and calls SDL_main */
+extern "C"
+int SDL_RunApp(int argc, char *argv[], SDL_main_func mainFunction, void *)
+{
+    int result;
+    HRESULT hr;
 
-        XTaskQueueCloseHandle(taskQueue);
+    hr = XGameRuntimeInitialize();
+    if (SUCCEEDED(hr)) {
+        XTaskQueueHandle taskQueue;
+        if (SDL_GetGDKTaskQueue(&taskQueue)) {
+            XTaskQueueSetCurrentProcessTaskQueue(taskQueue);
+
+            // Try to get the TitleID and initialize Xbox Live
+            Uint32 titleId;
+            bool xblInitialized = false;
+            hr = XGameGetXboxTitleId(&titleId);
+            if (SUCCEEDED(hr)) {
+                XblInitArgs xblInitArgs;
+                char scidBuffer[64];
+                SDL_zero(xblInitArgs);
+                xblInitArgs.queue = taskQueue;
+                SDL_snprintf(scidBuffer, 64, "00000000-0000-0000-0000-0000%08X", titleId);
+                xblInitArgs.scid = scidBuffer;
+                hr = XblInitialize(&xblInitArgs);
+                if (SUCCEEDED(hr)) {
+                    xblInitialized = true;
+                } else {
+                    SDL_SetError("[GDK] Unable to call XblInitialize");
+                }
+            } else {
+                SDL_SetError("[GDK] Unable to get TitleID. Will not call XblInitialize. Check MicrosoftGame.config!");
+            }
+
+            if (GDK_RegisterChangeNotifications()) {
+                // We are now ready to call the main function.
+                SDL_SetMainReady();
```

Comment:
I don't think it matters but I'll just quickly note that I swapped the order of `GDK_RegisterChangeNotifications()` and `SDL_SetMainReady()` around.